### PR TITLE
Fix `decide_worker_rootish_queuing_disabled` assert

### LINKED ISSUE #7063: `decide_worker_rootish_queuing_disabled` assertion fails when retiring worker
While stress-testing https://github.com/dask/distributed/pull/7062, `test_RetireWorker_stress`, which gracefully removes the best part of the cluster while performing a very heavy computation, failed once out of 162 runs:

https://github.com/crusaderky/distributed/actions/runs/3114670981/jobs/5050785452#step:18:1674

```
2022-09-23 18:56:03,193 - distributed.scheduler - ERROR - (<WorkerState 'tcp://127.0.0.1:63881', name: 6, status: closing_gracefully, memory: 21, processing: 27>, {<WorkerState 'tcp://127.0.0.1:63869', name: 0, status: running, memory: 61, processing: 6>, <WorkerState 'tcp://127.0.0.1:63879', name: 5, status: running, memory: 59, processing: 14>, <WorkerState 'tcp://127.0.0.1:63885', name: 8, status: running, memory: 59, processing: 17>, <WorkerState 'tcp://127.0.0.1:63877', name: 4, status: running, memory: 58, processing: 5>, <WorkerState 'tcp://127.0.0.1:63887', name: 9, status: running, memory: 59, processing: 6>})

Traceback (most recent call last):

  File "d:\a\distributed\distributed\distributed\scheduler.py", line 2040, in transition_waiting_processing

    if not (ws := self.decide_worker_rootish_queuing_disabled(ts)):

  File "d:\a\distributed\distributed\distributed\scheduler.py", line 1901, in decide_worker_rootish_queuing_disabled

    assert ws in self.running, (ws, self.running)
```

### PR DESCRIPTION
**Pull request by gjoseph92**
Closes #7063

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by crusaderky** [comment]
Is it possible to write a unit test that deterministically reproduces the problem? As of #7062, the failure rate is 1/162.

**Comment by fjetter** [comment]
@gjoseph92 what's the status on this?

**Comment by crusaderky** [comment]
@gjoseph92 any update on this?

**Comment by gjoseph92** [comment]
I'm finding it pretty hard to come up with a test for this. I imagine this case is happening when `TaskGroup.last_worker` is set to a worker that has just been put in `closing_gracefully` state, and `TaskGroup.last_worker_tasks_left` is non-zero. We go through a transitions cycle and pick this non-running last-used worker.

But I don't understand how this could actually happen.

A `transitions` cycle holds control of the event loop until it's done. So I don't think it's possible for a worker to switch from running to non-running in between two transitions.

And this line is intended to clear `last_worker` at the end of each transitions cycle: https://github.com/dask/distributed/blob/09837312ccb3a9f1a14ea068ba5825963fec82cc/distributed/scheduler.py#L2114-L2116

My guess is that there's some edge case that that logic isn't catching. Generally though, if this rootish task is going waiting->processing, and other tasks in the group are currently released or waiting, I'd expect those tasks to also be recommended waiting->processing in the same transitions cycle. Then the last task would clear out `last_worker`.

It seems there's a case where that doesn't happen, so `last_worker` doesn't get cleared at the end of one transitions cycle, and then we try to reuse it in a different cycle after the worker has switched to `closing_gracefully`.

Bottom line is that I'm reluctant to spend more time trying to come up with a test for this edge case, especially when we're looking to move away from this code path anyway by turning queuing on by default, and by using a more sensible (and stateless) algorithm for co-assignment https://github.com/dask/distributed/pull/7141.

Instead, I've switched this to just not re-use a previous worker if it's not running, which seems reasonable enough.

@fjetter @crusaderky how do you feel about merging this without a specific test?

**Comment by crusaderky** [comment]
> A `transitions` cycle holds control of the event loop until it's done. So I don't think it's possible for a worker to switch from running to non-running in between two transitions.

This is correct.

> My guess is that there's some edge case that that logic isn't catching.

I think this is what is happening:

1. tasks (x, 0), (x, 1), (x, 2), (x, 3) transition to processing. tg.last_worker is set.
2. The worker where (x, 2) is processing transitions to closing_gracefully, e.g. not running.
3. Work stealing steals (x, 2) (by mere coincidence). This happens before the worker is still flushing out other tasks in memory and has not completely left the cluster.
4. task (x, 2) transitions back to released, which in turn kicks off another transition to processing
5. tg.last_worker from the previous `transitions` loop is still there.
 
> @fjetter @crusaderky how do you feel about merging this without a specific test?

Personally I am quite unhappy to have half-dead, poorly tested code paths.

**Comment by gjoseph92** [comment]
> 5\. tg.last_worker from the previous `transitions` loop is still there

I like the work-stealing theory. But more fundamentally, how could `last_worker` be there from the last transitions loop? If the last transitions loop ran to completion, then you'd expect there to be exactly 1 released or waiting task in the last iteration of the loop, which would have cleared `last_worker`.

**Comment by crusaderky** [comment]
The good news:
Work stealing was innocent. I pushed a deterministic reproducer.

The bad news:
The reproducer highlights a failure when queueing is active.

File: distributed/tests/test_scheduler.py [hunk 1]
```diff
@@ -253,6 +254,72 @@ def random(**kwargs):
...
+        while a.state.executing_count != 1 or b.state.executing_count != 1:
+            await asyncio.sleep(0.01)
+
+        # - y-2 has no restrictions
+        # - TaskGroup(y) has more than 4 tasks (total_nthreads * 2)
+        # - TaskGroup(y) has less than 5 dependency groups
+        # - TaskGroup(y) has less than 5 dependency tasks
+        assert s.is_rootish(s.tasks["y-2"])
+
+        await evx[0].set()
```
Comment (gjoseph92) [inline distributed/tests/test_scheduler.py:296]: Okay, this is basically the issue with https://github.com/dask/distributed/blob/09837312ccb3a9f1a14ea068ba5825963fec82cc/distributed/scheduler.py#L2113-L2116

The silly assumption this is making is effectively that all tasks in the TaskGroup have the same dependencies, so they're all runnable (and will all transition) at the same time. It's assuming that we're not going to have a subset of the group transition, then later transition another subset.

This is obviously not true, it just happens to pretty much always be the case in "common workloads" built with dask collections.

I've always really, really disliked the use of TaskGroups in the co-location logic in order to guess at things about graph structure. In this test, if you just change the _names_ of some the `y` tasks so they're in different groups, it'll probably pass. Task names should be opaque to dask and should not affect scheduling behavior. The combination of name-based heuristics with statefulness in the current implementation is even more problematic, and makes bugs like this possible.

To me, all this is just another argument for https://github.com/dask/distributed/pull/7141. A static way of determining these groupings based purely on graph structure would eliminate the possibility of these kinds of bugs.

Furthermore, in the graph you've constructed, the `y` tasks shouldn't even be treated as one co-assignment group. There should be three groups. The fact that there's only one group is just an artifact of the naming scheme (which totally can happen in real usage).

**Comment by gjoseph92** [comment]
@crusaderky your new test is passing for me; that's expected right? It seems like the `and lws.status == Status.running` check is solving the specific issue here, right?

**Comment by crusaderky** [comment]
> 



> @crusaderky your new test is passing for me; that's expected right? It seems like the `and lws.status == Status.running` check is solving the specific issue here, right?

Yes, my new test fails on main and passes on your branch with no_queue.
However, it fails with queue (#7204).

**Comment by gjoseph92** [comment]
Sounds good, I'm looking into #7204 now.

So this is more or less ready to go in that it resolves https://github.com/dask/distributed/issues/7063. We just can't merge it yet because the test fails on queuing CI due to #7204.

**Comment by gjoseph92** [comment]
This seems to be actually (intermittently) failing `distributed/tests/test_scheduler.py::test_include_communication_in_occupancy` on macOS, looking into it.

Also maybe `distributed/diagnostics/tests/test_task_stream.py::test_client_sync`. I can't reproduce that locally though.

Also all the normal websockets stuff and https://github.com/dask/distributed/issues/7208.

**Comment by gjoseph92** [comment]
Including https://github.com/dask/distributed/pull/7212 in here to get CI more greenish, but let's merge that PR before this one.

**Review by crusaderky — APPROVED** [review]
